pkg/agent/openflow/pipeline.go

```diff
@@ -1531,33 +1576,57 @@ func (c *client) serviceLBFlow(groupID binding.GroupIDType, svcIP net.IP, svcPor
 // to the Endpoint IP according to the Endpoint selection decision which is stored
 // in regs.
 func (c *client) endpointDNATFlow(endpointIP net.IP, endpointPort uint16, protocol binding.Protocol) binding.Flow {
-	ipVal := binary.BigEndian.Uint32(endpointIP)
+	ipProtocol := getIPProtocol(endpointIP)
 	unionVal := (marksRegServiceSelected << endpointPortRegRange.Length()) + uint32(endpointPort)
 	table := c.pipeline[endpointDNATTable]
-	return table.BuildFlow(priorityNormal).
-		Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
-		MatchProtocol(protocol).
-		MatchReg(int(endpointIPReg), ipVal).
-		MatchRegRange(int(endpointPortReg), unionVal, binding.Range{0, 18}).
-		Action().CT(true, table.GetNext(), CtZone).
-		DNAT(
-			&binding.IPRange{StartIP: endpointIP, EndIP: endpointIP},
-			&binding.PortRange{StartPort: endpointPort, EndPort: endpointPort},
-		).
-		LoadToMark(serviceCTMark).
-		CTDone().
-		Done()
+	if ipProtocol == binding.ProtocolIP {
```

Comment:
Since most parts of the flow are the same, what about change to this

```
flowBuilder := table.BuildFlow(priorityNormal).
		Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
		MatchRegRange(int(endpointPortReg), unionVal, binding.Range{0, 18}).
		MatchProtocol(ipProtocol)
if ipProtocol == binding.ProtocolIP {
    ipVal := binary.BigEndian.Uint32(endpointIP.To4())
    flowBuilder = flowBuilder.MatchReg(int(endpointIPReg), ipVal).
		MatchRegRange(int(endpointPortReg), unionVal, binding.Range{0, 18})
} else {
    ipVal := []byte(endpointIP)
    flowBuilder = flowBuilder.MatchXxReg(int(endpointIPv6XXReg), ipVal).
		MatchRegRange(int(endpointPortReg), unionVal, binding.Range{0, 18}).
}
return flowBuilder.Action().CT(true, table.GetNext(), CtZone).
		DNAT(
				&binding.IPRange{StartIP: endpointIP, EndIP: endpointIP},
				&binding.PortRange{StartPort: endpointPort, EndPort: endpointPort},
		).
		LoadToMark(serviceCTMark).
		CTDone().
		Done()
```

Reply:
Indeed much cleaner.